File: packages/relay-compiler/transforms/query-generators/ViewerQueryGenerator.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

// flowlint ambiguous-object-type:error

'use strict';

const {createUserError} = require('../../core/CompilerError');
const {
  buildFragmentSpread,
  buildOperationArgumentDefinitions,
} = require('./utils');

import type {Fragment} from '../../core/IR';
import type {Schema} from '../../core/Schema';
import type {QueryGenerator, RefetchRoot} from '.';

const VIEWER_TYPE_NAME = 'Viewer';
const VIEWER_FIELD_NAME = 'viewer';

function buildRefetchOperation(
  schema: Schema,
  fragment: Fragment,
  queryName: string,
): ?RefetchRoot {
  if (schema.getTypeString(fragment.type) !== VIEWER_TYPE_NAME) {
    return null;
  }
  // Handle fragments on viewer
  const queryType = schema.expectQueryType();
  const viewerType = schema.getTypeFromString(VIEWER_TYPE_NAME);
  const viewerField = schema.getFieldConfig(
    schema.expectField(queryType, VIEWER_FIELD_NAME),
  );
  const viewerFieldType = schema.getNullableType(viewerField.type);
  if (
    !(
      viewerType &&
      schema.isObject(viewerType) &&
      schema.isObject(viewerFieldType) &&
      schema.areEqualTypes(viewerFieldType, viewerType) &&
      viewerField.args.length === 0 &&
      schema.areEqualTypes(fragment.type, viewerType)
    )
  ) {
    throw createUserError(
      `Invalid use of @refetchable on fragment '${fragment.name}', check ` +
        "that your schema defines a 'Viewer' object type and has a " +
        "'viewer: Viewer' field on the query type.",
      [fragment.loc],
    );
  }
  return {
    identifierField: null,
    path: [VIEWER_FIELD_NAME],
    node: {
      argumentDefinitions: buildOperationArgumentDefinitions(
        fragment.argumentDefinitions,
      ),
      directives: [],
      kind: 'Root',
      loc: {kind: 'Derived', source: fragment.loc},
      metadata: null,
      name: queryName,
      operation: 'query',
      selections: [
        {
          alias: VIEWER_FIELD_NAME,
          args: [],
          connection: false,
          directives: [],
          handles: null,
          kind: 'LinkedField',
          loc: {kind: 'Derived', source: fragment.loc},
          metadata: null,
          name: VIEWER_FIELD_NAME,
          selections: [buildFragmentSpread(fragment)],
          type: schema.assertLinkedFieldType(viewerField.type),
        },
      ],
      type: queryType,
    },
    transformedFragment: fragment,
  };
}

module.exports = ({
  description: 'the Viewer type',
  buildRefetchOperation,
}: QueryGenerator);

File: compiler/crates/relay-transforms/src/refetchable_fragment/viewer_query_generator.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::sync::Arc;

use common::Diagnostic;
use common::DiagnosticsResult;
use common::WithLocation;
use graphql_ir::FragmentDefinition;
use graphql_ir::LinkedField;
use graphql_ir::OperationDefinitionName;
use graphql_ir::Selection;
use relay_config::SchemaConfig;
use schema::FieldID;
use schema::SDLSchema;
use schema::Schema;
use schema::Type;

use super::CONSTANTS;
use super::QueryGenerator;
use super::RefetchRoot;
use super::RefetchableMetadata;
use super::build_fragment_metadata_as_directive;
use super::build_fragment_spread;
use super::build_operation_variable_definitions;
use super::build_used_global_variables;
use super::uses_prefetchable_pagination_in_connection;
use super::validation_message::ValidationMessage;
use crate::root_variables::VariableMap;

fn build_refetch_operation(
    schema: &SDLSchema,
    _schema_config: &SchemaConfig,
    fragment: &Arc<FragmentDefinition>,
    query_name: OperationDefinitionName,
    variables_map: &VariableMap,
) -> DiagnosticsResult<Option<RefetchRoot>> {
    if schema.get_type_name(fragment.type_condition) != CONSTANTS.viewer_type_name {
        return Ok(None);
    }
    let query_type = schema.query_type().unwrap();
    let viewer_field_id = get_viewer_field_id(schema, query_type, fragment)?;

    let fragment = Arc::new(FragmentDefinition {
        directives: build_fragment_metadata_as_directive(
            fragment,
            RefetchableMetadata {
                operation_name: query_name,
                path: vec![CONSTANTS.viewer_field_name],
                identifier_info: None,
                is_prefetchable_pagination: uses_prefetchable_pagination_in_connection(fragment),
            },
        ),
        used_global_variables: build_used_global_variables(
            variables_map,
            &fragment.variable_definitions,
        )?,
        variable_definitions: fragment.variable_definitions.clone(),
        ..fragment.as_ref().clone()
    });
    Ok(Some(RefetchRoot {
        variable_definitions: build_operation_variable_definitions(&fragment),
        selections: vec![Selection::LinkedField(Arc::new(LinkedField {
            alias: None,
            definition: WithLocation::new(fragment.name.location, viewer_field_id),
            arguments: vec![],
            directives: vec![],
            selections: vec![build_fragment_spread(&fragment)],
        }))],
        fragment,
    }))
}

fn get_viewer_field_id(
    schema: &SDLSchema,
    query_type: Type,
    fragment: &FragmentDefinition,
) -> DiagnosticsResult<FieldID> {
    let viewer_type = schema.get_type(CONSTANTS.viewer_type_name);
    let viewer_field_id = schema.named_field(query_type, CONSTANTS.viewer_field_name);
    if let Some(viewer_type) = viewer_type
        && let Some(viewer_field_id) = viewer_field_id
    {
        let viewer_field = schema.field(viewer_field_id);
        if viewer_type.is_object()
            && viewer_type == viewer_field.type_.inner()
            && viewer_type == fragment.type_condition
            && viewer_field.arguments.is_empty()
        {
            return Ok(viewer_field_id);
        }
    }
    Err(vec![Diagnostic::error(
        ValidationMessage::InvalidViewerSchemaForRefetchableFragmentOnViewer {
            fragment_name: fragment.name.item,
        },
        fragment.name.location,
    )])
}

pub const VIEWER_QUERY_GENERATOR: QueryGenerator = QueryGenerator {
    description: "the Viewer type",
    build_refetch_operation,
};
